Implement CoinJoin signing flow

andrewkozlik opened the pull request:
This is a draft of a solution to #1053. There are still several TODO items, so at this point just go over the high-level design, focusing on core/src/apps/bitcoin/preauthorization.py. We can then discuss the items below.

TODO:
1. The `SignTx` CoinJoin message should be enabled also in soft-locked mode.
2. The `AuthorizeCoinJoin` message should cache the account node in the `CoinJoinAuthorization` object, so that it can be used by `GetOwnershipProof` and `SignTx` when the device is soft-locked. Not necessary yet, because we are still keeping the seed in memory unencrypted. But we should at least consider how this will be done.
3. Agree on what exactly should be checked by the firmware and whether to fail or prompt if fees are over threshold.
4. Consider an alternative design: Instead of deriving the `CoinJoin` class from `Bitcoin`, make `CoinJoin` a strategy of `Bitcoin`. This would easily allow CoinJoin for Bitcoinlike coins.

matejcik commented on core/src/apps/bitcoin/authorize_coinjoin.py:
```diff
@@ -0,0 +1,63 @@
+    text.bold("%s %s" % (format_amount(msg.max_fee, coin.decimals), coin.coin_shortcut))
+    await require_hold_to_confirm(ctx, text)
+
+    set_coinjoin_authorization(msg, coin)
```
Worth considering: overwriting previous authorizations should not be done silently.

matejcik commented on core/src/apps/bitcoin/authorize_coinjoin.py:
```diff
@@ -0,0 +1,63 @@
+
+    set_coinjoin_authorization(msg, coin)
+
+    return Success
```
typo: `return Success(message="something")`

andrewkozlik commented:
> What stands out to me is that preauthorization allows the `user_confirmation` bit. Is it really OK to do that? Shouldn't unattended ownership proofs automatically be not-user-confirmed?

The user authorized the creation of ownership proofs for `msg.coordinator`, and the proof will only be automatically signed with the `user_confirmation` flag if the `commitment_data` starts with `coordinator` followed by exactly 8 bytes of round ID. So the user preauthorized creating proofs which respect this commitment data format. BTW, I am thinking about increasing the round ID to 32 bytes or possibly even wrapping this into some data structure which would allow variable-length round IDs.

The purpose of the `user_confirmation` flag is to assure the server that the owner of UTXOs expressed some sort of consent to participate in a CoinJoin with that server, acting as a defense against DoS attacks.

andrewkozlik commented:
> The `preauthorization` module LGTM at the high level. A little unsure about the duplication of methods/functions, but let's look into that in more detail later.

If I understand correctly, you are referring to the fact that there are pairs of functions like `confirm_get_ownership_proof()` and `CoinJoinAuthorization::confirm_get_ownership_proof()`. One motivation behind that is that in time we might want to autoconfirm for example an ownership proof for various reasons (some other feature than CoinJoin) or we might want to support multiple CoinJoin preauthorizations. In that case `confirm_get_ownership_proof()` takes care of the dispatch. Maybe it's premature.

matejcik commented:
> If I understand correctly, you are referring to the fact that there are pairs of functions like `confirm_get_ownership_proof()` and `CoinJoinAuthorization::confirm_get_ownership_proof()`. One motivation behind that is that in time we might want to autoconfirm for example an ownership proof for various reasons (some other feature than CoinJoin) or we might want to support multiple CoinJoin preauthorizations. In that case `confirm_get_ownership_proof()` takes care of the dispatch. Maybe it's premature.

i understand the rationale, but i'm thinking "there must be a way to make the implementation nicer" :)

andrewkozlik commented:
> 4\. Consider an alternative design: Instead of deriving the `CoinJoin` class from `Bitcoin`, make `CoinJoin` a strategy of `Bitcoin`. This would easily allow CoinJoin for Bitcoinlike coins.

@matejcik, in a separate [branch](https://github.com/trezor/trezor-firmware/compare/andrewkozlik/coinjoin-strategy) I implemented an `Authorizer` and `CoinJoinAuthorizer` class to handle the computation of the total amounts and requesting/verifying user confirmations, see https://github.com/trezor/trezor-firmware/commit/a160c5dee0d0752c2a7f74a09b4853f30d47b053. Let me know what you think of this approach. I think I like it a bit better than the present one. Also thinking of merging `self.step2_confirm_outputs()` and `step3_confirm_tx()`.

matejcik commented:
> @matejcik, in a separate [branch](https://github.com/trezor/trezor-firmware/compare/andrewkozlik/coinjoin-strategy) I implemented an `Authorizer` and `CoinJoinAuthorizer` class to handle the computation of the total amounts and requesting/verifying user confirmations, see [a160c5d](https://github.com/trezor/trezor-firmware/commit/a160c5dee0d0752c2a7f74a09b4853f30d47b053). Let me know what you think of this approach.

I like it a lot.

BTW, it makes it clearer that the `preauthorization` object should be fetched from cache in the protobuf handler and passed to the `Authorizer` instance.


> Also thinking of merging `self.step2_confirm_outputs()` and `step3_confirm_tx()`.

more like, calling the `authorizer` from `signer()` directly, instead of having a separate step?

matejcik commented:
quick draft of the `Preauthorized` handler/dispatcher:

* `apps.preauthorized` namespace
* `preauthorize(message_type, preauthorization_object)` can keep or evict other preauth types, and/or show UI information
* `preauthorized` handler will fetch the current preauthorization object and send `ctx.call(PreauthRequest, message_type)`. This guarantees that the result is of the preauthorized type.
* the preauth handler can query the main dispatcher for the function to call, or it can be registered via the `preauthorize()` function as well

andrewkozlik commented on core/src/apps/bitcoin/sign_tx/__init__.py:
```diff
@@ -2,41 +2,55 @@
+    authorization: Optional[CoinJoinAuthorization] = None,
 ) -> TxRequest:
-    if not utils.BITCOIN_ONLY:
+    from . import approvers, bitcoin, helpers, layout, progress
+    from trezor.messages.TxRequest import TxRequest
```
Something fishy is going on here. Looks like `bitcoin` and some other modules are getting unimported.

andrewkozlik commented:
> quick draft of the `Preauthorized` handler/dispatcher:
> 
>     * `apps.preauthorized` namespace
> 
>     * `preauthorize(message_type, preauthorization_object)` can keep or evict other preauth types, and/or show UI information
> 
>     * `preauthorized` handler will fetch the current preauthorization object and send `ctx.call(PreauthRequest, message_type)`. This guarantees that the result is of the preauthorized type.
> 
>     * the preauth handler can query the main dispatcher for the function to call, or it can be registered via the `preauthorize()` function as well

See https://github.com/trezor/trezor-firmware/pull/1094/commits/ec45e424c54285dc3871934af0db224721819093 and https://github.com/trezor/trezor-firmware/pull/1094/commits/ffc744c1a31e881d1d5075e4022e9746686a4bd9. I wrote it so that there is only one authorization object per session and that object also handles the dispatch.

andrewkozlik commented:
After becoming more acquainted with how the module importing/unimporting works I reworked the dispatch once more to have the handling be done by querying the main dispatcher. I also changed `@with_keychain` to get the keychain from an optional authorization object if it's provided. The core of these changes can be seen in https://github.com/trezor/trezor-firmware/pull/1094/commits/cf822f53ba04eb40321fd1ef7f5144efff48d535 and https://github.com/trezor/trezor-firmware/pull/1094/commits/9fd9b4a81db1fc67e4931d6194ec457ff759e682, respectively.

Having an optional authorization object in `@with_keychain` nicely simplifies some things but creates a bit of a problem in typing, because `find_registered_workflow_handler()` returns a handler which should be taking just two arguments, not two + one optional: https://github.com/trezor/trezor-firmware/pull/1094/commits/cf822f53ba04eb40321fd1ef7f5144efff48d535#diff-e0f057bf56465baa0eb87e69b50267e8R145. However, the typing inside `find_registered_workflow_handler()` seems flaky to begin with, because `import_workflow()` returns `Any`.

@matejcik If you agree with the high-level design, then we can proceed to a standard code review. For that I will squash the fixups, reword some commits and open a fresh PR.

matejcik commented:
From a very quick overview LGTM. The `DoPreauthorized` handler implementation looks good.
I previously ACKed the high-level design of the preauthorization handling, and also the strategy-like Approvers.

I'm not sure if there's anything else I should be checking at this stage? So if that's it please go ahead with the full PR

andrewkozlik commented:
Squashed, rebased and opened a new PR: https://github.com/trezor/trezor-firmware/pull/1127.
